### ISSUE TITLE
sbt 1.0.2 vs 0.13.16: different test TaskDefs

### ISSUE THREAD

**Issue by cunei** [sbt/sbt#3583]
### steps
```
$ cat build.sbt 
version := "0.0.1"
name := "broken"
organization := "org.catastrophe"
scalaVersion := "2.10.6"
libraryDependencies += "org.scalacheck" %% "scalacheck" % "1.13.5" % "test"
$ cat ./src/test/scala/q/X.scala
package q
object X {
  object Y extends org.scalacheck.Properties("a")
}
$ cat project/build.properties 
sbt.version=1.0.2
#sbt.version=0.13.16
```
## problem
`sbt test` succeeds with 0.13.16, but with 1.0.2 it says:
```
[error] java.lang.ClassNotFoundException: q.X.Y$
[error]         at java.lang.ClassLoader.findClass(ClassLoader.java:530)
[error]         at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[error]         at sbt.internal.inc.classpath.ClasspathFilter.loadClass(ClassLoaders.scala:70)
[error]         at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[error]         at java.lang.Class.forName0(Native Method)
[error]         at java.lang.Class.forName(Class.java:348)
[error]         at org.scalacheck.Platform$.loadModule(Platform.scala:67)
[error]         at org.scalacheck.ScalaCheckRunner$BaseTask.<init>(ScalaCheckFramework.scala:49)
[error]         at org.scalacheck.ScalaCheckRunner$$anon$1.<init>(ScalaCheckFramework.scala:81)
[error]         at org.scalacheck.ScalaCheckRunner.rootTask(ScalaCheckFramework.scala:81)
```
### notes
I dug into the Scalacheck code. The code receives a bunch of `sbt.testing.TaskDef` instances (as per sbt's `test-interface`), and tries a `Class.forName(name + "$",...` on each. The list does not contain `q.X.Y`, above in 0.13.16, but it does in 1.0.2, and the `forName` fails.

0.13.16:
```
$ sbt 'show testGrouping'
[info] * Group(<default>,List(),InProcess)
```
1.0.2:
```
$ sbt 'show testGrouping'
[info] * Group(<default>,Vector(Test q.X.Y : subclass(true, org.scalacheck.Properties)),InProcess)
```
Any comments on the discrepancy? I would assume that the same scalaVersion and libraries should lead to the same testing behavior (except for intentional fixes).

**Comment by jvican** [comment]
This could be a problem coming from Zinc analysis, which is then exposed via sbt's test-interface.

**Comment by eed3si9n** [comment]
Maybe there are two issues here:

1. Is sbt 1 right in detecting `q.X.Y` as a test group?
2. Is the name `q.X.Y` correct?

One thing that could be happening is that we improved in detecting nested object, but we messed up in calculating the ["canonical" name](https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.7) of object-nested-in-object. If in fact the requirement is that it must support `Class.forName(...)` that's an interesting requirement that could clarify the discussion.

**Comment by jvican** [comment]
Gregorz made the switch to canonical names. I want to believe that for a good reason. I'll be resuming my work on Zinc soon, I can look at that issue to get warmed up then.
  
All I can say for now is that `q.X.Y$` is a valid class name in this context, and should exist.

**Comment by cunei** [comment]
So, the matter seems to be not too complicated, even though solving it requires some thinking. A TaskDef contains a fully qualified name. The nested object `q.X.Y` is actually encoded as `q/X$Y$.class`. That means that the forName should be applied to `q.X$Y$`, rather than to `q.X.Y$`. The issue is that a `TaskDef` refers to the contained object using the fullyQualifiedName `q.X.Y`, which is indistinguishable from an object `Y` in a package `q.X`, so getting back to the class file by using `forName()` is tricky when nested objects are involved.
The possible approaches that I can see are:
1) Do not create a TaskDef for nested objects as apparently was the case in 0.13, therefore restoring the previous behavior
2) Extend TaskDef, adding a method that can calculate a class name that `forName()` can understand in any case.
3) Tweak Fingerprint, distinguishing whether nested objects should be included or not. Currently `SubclassFingerprint` has a flag to distinguish whether modules (objects) should be included or not, for example. The meaning of `SubclassFingerprint` could be reverted to the 0.13 behavior, excluding nested objects, while a new Fingerprint (a subclass of `SubclassFingerprint`) could be used to include the nested objects and classes as well, while also including a classfile name conversion method.

If that is ok, I will look into filtering out the nested objects as per option 1 above, and subsequently implementing option 3 as a possible extension.

**Comment by dwijnand** [comment]
Fixed in #3669.
